### PR TITLE
feature: domain resolver

### PR DESCRIPTION
**Pull request by banklesss**
Working with unstoppable and $handle domains at the moment

Related to [YOMO-898](https://emurgo.atlassian.net/browse/YOMO-898)

TODO:

- [x] add test
- [x] add proper typing
- [x] refactor folder structure
- [ ] add cns resolver
- [x] add $handle resolver
- [x] add unstoppable domains resolver
- [x] use fetcher instead of axios directly
- [x] improve error management
- [x] UI update
- [x] React Context Provider
- [x] Initial message storage API
- [ ] ~~preprod support~~

[YOMO-898]: https://emurgo.atlassian.net/browse/YOMO-898?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/wallet-mobile/src/TxHistory/TxHistoryNavigator.tsx [hunk 1]
```diff
@@ -97,218 +107,220 @@ export const TxHistoryNavigator = () => {
...
-                  title: strings.editAmountTitle,
-                  ...sendOptions,
+        <SwapFormProvider>
+          <ResolverProvider resolverModule={resolverModule}>
+            <ClaimProvider key={wallet.id} claimApi={claimApi}>
+              <Stack.Navigator
+                screenListeners={{}}
+                screenOptions={{
+                  ...defaultStackNavigationOptions,
+                  detachPreviousScreen: false /* https://github.com/react-navigation/react-navigation/issues/9883 */,
```
Comment (michalszorad) [inline apps/wallet-mobile/src/TxHistory/TxHistoryNavigator.tsx:123]: Was this change intentional?

File: apps/wallet-mobile/src/components/Icon/Check.tsx [hunk 2]
```diff
@@ -2,21 +2,19 @@ import React from 'react'
...
-
 type Props = {
   size?: number
   color?: string
   style?: ImageStyle
 }
 
-export const Check = ({size = 40, color}: Props) => (
-  <Svg width={size} height={size} viewBox="-2 -2 28 28">
+export const Check = ({size = 40, color = 'black'}: Props) => (
```
Comment (michalszorad) [inline apps/wallet-mobile/src/components/Icon/Check.tsx:11]: Do we need to check all the places where is it used?

File: apps/wallet-mobile/src/features/Send/useCases/StartMultiTokenTx/InputReceiver/InputReceiver.tsx [hunk 3]
```diff
@@ -29,6 +42,57 @@ export const InputReceiver = ({isLoading, ...props}: {isLoading: boolean} & Text
...
+      Animated.timing(spin, {
+        toValue: 1,
+        duration: 1500,
+        easing: Easing.linear,
+        useNativeDriver: true,
+      }),
+    ).start()
+  }, [spin])
+
+  const getRotationStyle = () => {
```
Comment (michalszorad) [inline apps/wallet-mobile/src/features/Send/useCases/StartMultiTokenTx/InputReceiver/InputReceiver.tsx]: Maybe this fn can be extracted out of the component

File: apps/wallet-mobile/src/features/Send/useCases/StartMultiTokenTx/MultiAddressResolutionNotice/MultiAddressResolutionNotice.tsx [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+import {Icon, Spacer} from '../../../../../components'
+
+export const MultiAddressResolutionNotice = () => {
+  return (
+    <View style={styles.notice}>
+      <Icon.Info size={30} color="#ECBA09" />
+
+      <Spacer height={8} />
+
+      <Text style={styles.text}>There are two addresses for this domain. Please choose the desired domain.</Text>
```
Comment (michalszorad) [inline apps/wallet-mobile/src/features/Send/useCases/StartMultiTokenTx/MultiAddressResolutionNotice/MultiAddressResolutionNotice.tsx]: This may need a translation

File: apps/wallet-mobile/src/features/Send/useCases/StartMultiTokenTx/StartMultiTokenTxScreen.tsx [hunk 5]
```diff
@@ -42,10 +62,59 @@ export const StartMultiTokenTxScreen = () => {
...
+    [addressValidationError, receiver, succesfulResolvedAddresses.length],
+  )
+
+  const {isLoading: isResolutionLoading} = useResolverAddresses(
+    {receiver},
+    {
+      onSettled(addresses) {
+        const succesfulResolvedAddresses = addresses?.filter(({address}) => address !== null) ?? []
+
+        if (succesfulResolvedAddresses?.length > 0) {
```
Comment (michalszorad) [inline apps/wallet-mobile/src/features/Send/useCases/StartMultiTokenTx/StartMultiTokenTxScreen.tsx]: Coud it resolve to more addresses? In that case we want to use only the 1st one?

File: packages/resolver/src/adapters/cns.test.ts [hunk 6]
```diff
@@ -0,0 +1,3 @@
+test('dummy', () => {
```
Comment (michalszorad) [inline packages/resolver/src/adapters/cns.test.ts]: This may not be needed anymore 😄

File: packages/resolver/src/adapters/handle-api.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+export const getHandleCryptoAddress = async (
+  receiverDomain: Resolver.Receiver['domain'],
+): Promise<string> => {
+  try {
+    const validatedHandle = stringWithDollarSchema.parse(receiverDomain)
+    const handle = validatedHandle.replace(/^\$/, '')
+
+    const config = {
+      method: 'get',
+      url: `https://api.handle.me/handles/${handle}`,
```
Comment (michalszorad) [inline packages/resolver/src/adapters/handle-api.ts]: Does it work on mainnet and on testnet as well?

File: apps/wallet-mobile/package.json [hunk 8]
```diff
@@ -262,7 +262,7 @@
     "prettier-plugin-packagejson": "^2.2.11",
     "prettylint": "^1.0.0",
     "react-addons-test-utils": "^15.6.2",
-    "react-devtools-core": "4.26.1",
+    "react-devtools-core": "^4.28",
```
Comment (stackchain) [inline apps/wallet-mobile/package.json:266]: had to bump it, it wasn't working when selecting another context

File: apps/wallet-mobile/src/TxHistory/TxHistoryNavigator.tsx (hunk 1 above)
Comment (stackchain) [reply apps/wallet-mobile/src/TxHistory/TxHistoryNavigator.tsx:123]: reverted

File: apps/wallet-mobile/src/components/Icon/Check.tsx (hunk 2 above)
Comment (stackchain) [reply apps/wallet-mobile/src/components/Icon/Check.tsx:11]: it was fall-backing to black already on line 19.

File: apps/wallet-mobile/src/components/TextInput/TextInput.tsx
Comment (stackchain) [inline apps/wallet-mobile/src/components/TextInput/TextInput.tsx:1]: this component needs to be refactored.

File: apps/wallet-mobile/src/features/Send/common/SendContext.tsx [hunk 9]
```diff
@@ -137,74 +154,96 @@ export type TargetAction =
...
-        }
+      case 'addressRecordsFetched': {
+        const {addressRecords} = action
+        const selectedTargetIndex = state.selectedTargetIndex
+
+        draft.forEach((target, index) => {
+          if (index === selectedTargetIndex) {
+            if (addressRecords !== undefined) {
+              const keys = Object.keys(addressRecords).filter(isNameServer)
+              const nameServer = keys.length === 1 ? keys[0] : undefined
```
Comment (stackchain) [inline apps/wallet-mobile/src/features/Send/common/SendContext.tsx:188]: when only 1 record is found, it is selected.

File: apps/wallet-mobile/src/features/Send/common/useSendReceiver.tsx [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+  )
+  const debouncedRefetch = React.useMemo(() => debounceMaker(refetch, 300), [refetch])
+
+  const cancelPendingRequests = React.useCallback(
+    () => queryClient.cancelQueries({queryKey: ['useResolverCryptoAddresses']}),
+    [queryClient],
+  )
+
+  React.useEffect(() => {
+    if (receiver.as === 'domain') cancelPendingRequests().then(() => debouncedRefetch.call())
```
Comment (stackchain) [inline apps/wallet-mobile/src/features/Send/common/useSendReceiver.tsx:36]: if is domain/address abort sign abort pending requests, CNS module will have to expose it.

File: apps/wallet-mobile/src/features/Send/useCases/ConfirmTx/Summary/ReceiverInfo.tsx
Comment (stackchain) [inline apps/wallet-mobile/src/features/Send/useCases/ConfirmTx/Summary/ReceiverInfo.tsx:1]: tx summary will be reworked

File: apps/wallet-mobile/src/yoroi-wallets/cardano/constants/common.ts [hunk 11]
```diff
@@ -54,7 +54,7 @@ export const LINEAR_FEE = {
 } as const
 export const MINIMUM_UTXO_VAL = '1000000'
 
-export const HISTORY_REFRESH_TIME = 25000
+export const HISTORY_REFRESH_TIME = 35000
```
Comment (stackchain) [inline apps/wallet-mobile/src/yoroi-wallets/cardano/constants/common.ts:57]: avg block is ~20s, doing below 30s has a high chance to sync with no updates, ideal 30/worst 40/avg 35s. (suggesting to add as a config) for the user

File: apps/wallet-mobile/src/yoroi-wallets/hooks/index.ts [hunk 12]
```diff
@@ -742,7 +742,7 @@ export const useIsOnline = (
         () => true,
         () => false,
       ),
-    refetchInterval: 5000,
+    refetchInterval: 15000,
```
Comment (stackchain) [inline apps/wallet-mobile/src/yoroi-wallets/hooks/index.ts:736]: too often, under a block is fine avg 20s

File: packages/resolver/src/adapters/cns.ts [hunk 13]
```diff
@@ -0,0 +1,11 @@
+import {Resolver} from '@yoroi/types'
+
+export const getCnsCryptoAddress = async (receiver: string) => {
+  if (!receiver.includes('.')) throw new Resolver.Errors.InvalidDomain()
+  if (!isCnsDomain(receiver))
+    return Promise.reject(new Resolver.Errors.UnsupportedTld())
+  return Promise.reject(new Resolver.Errors.UnsupportedTld())
```
Comment (stackchain) [inline packages/resolver/src/adapters/cns.ts:7]: unsupported for now, needs implementation.

File: packages/resolver/src/adapters/handle-api.ts [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+  length: number
+  og_number: number
+  //   export enum Rarity {
+  //     basic = 'basic', // - 8-15 characters
+  //     common = 'common', // - 4-7 characters
+  //     rare = 'rare', // - 3 characters
+  //     ultra_rare = 'ultra_rare', // - 2 characters
+  //     legendary = 'legendary' // - 1 character
+  // }
+  rarity: string // translated to string only
```
Comment (stackchain) [inline packages/resolver/src/adapters/handle-api.ts:61]: rarity on handle API is an enum, we translate it to string for now.

File: packages/types/src/resolver/receiver.ts [hunk 15]
```diff
@@ -0,0 +1,8 @@
+import {ResolverNameServer} from './service'
+
+export type ResolverReceiver = {
```
Comment (stackchain) [inline packages/types/src/resolver/receiver.ts:3]: receiver is no longer a string and becomes a new data structure, 
`resolve` => domain/address/receiver to use/resolve aka URL
`as` => type, `domain` means supported domain
`selectedNameServer` defaults to the only response if === 1 aka NS 
`addressRecords` the list of addresses per name server aka A record

File: apps/wallet-mobile/src/features/Send/useCases/StartMultiTokenTx/InputReceiver/InputReceiver.tsx (hunk 3 above)
Comment (stackchain) [reply apps/wallet-mobile/src/features/Send/useCases/StartMultiTokenTx/InputReceiver/InputReceiver.tsx]: updated to use the current activity indicator.

File: apps/wallet-mobile/src/features/Send/useCases/StartMultiTokenTx/StartMultiTokenTxScreen.tsx (hunk 5 above)
Comment (stackchain) [reply apps/wallet-mobile/src/features/Send/useCases/StartMultiTokenTx/StartMultiTokenTxScreen.tsx]: it works differently now. if resolves 1, it's defaulted otherwise is expected the user input

File: apps/wallet-mobile/translations/messages/src/WalletInit/WalletForm.json [hunk 16]
```diff
@@ -66,12 +98,20 @@
     "start": {
       "line": 149,
       "column": 28,
+<<<<<<< HEAD
+      "index": 5304
+=======
```
Comment (michalszorad) [inline apps/wallet-mobile/translations/messages/src/WalletInit/WalletForm.json]: This looks like something went wrong during a merge

Comment (stackchain) [reply apps/wallet-mobile/translations/messages/src/WalletInit/WalletForm.json]: 👀

**Review by michalszorad — APPROVED** [review]
Tried and works well!

File: apps/wallet-mobile/translations/messages/src/WalletInit/WalletForm.json (hunk 16 above)
Comment (stackchain) [reply apps/wallet-mobile/translations/messages/src/WalletInit/WalletForm.json]: uber vscode, fixed.

File: packages/resolver/src/adapters/handle-api.ts [hunk 17]
```diff
@@ -0,0 +1,106 @@
+import {Api, Resolver} from '@yoroi/types'
+import {fetchData, FetchData, handleApiError, isLeft} from '@yoroi/common'
+import {z} from 'zod'
+import {AxiosRequestConfig} from 'axios'
```
Comment (banklesss) [inline packages/resolver/src/adapters/handle-api.ts:4]: Should it be wrapped in `@yoroi/common`?

File: packages/resolver/src/adapters/handle-api.ts [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+        )
+        return parsedResponse.resolved_addresses.ada
+      }
+    } catch (error: unknown) {
+      return handleHandleApiError(error)
+    }
+  }
+}
+
+// https://github.com/koralabs/handles-public-api/blob/master/src/models/view/handle.view.model.ts
```
Comment (banklesss) [inline packages/resolver/src/adapters/handle-api.ts]: i think this should refer to the current commit

Comment (stackchain) [reply packages/resolver/src/adapters/handle-api.ts]: that makes sense, can you add as a suggestion ?

File: packages/resolver/src/adapters/handle-api.ts (hunk 17 above)
Comment (stackchain) [reply packages/resolver/src/adapters/handle-api.ts:4]: yap, is a little bit leaky since is exposing the "fetcher" config directly, but also didn't want a 1:1, it's a @banklesss good call for an improvement, maybe we can expose only what we are actually using.

File: packages/resolver/src/adapters/handle-api.ts (hunk 18 above)
Comment (banklesss) [reply packages/resolver/src/adapters/handle-api.ts:44]: https://github.com/koralabs/handles-public-api/blob/fd9d9f2cf3143e317a780b81b22869a755ab6af8/src/models/view/handle.view.model.ts

Comment (banklesss) [inline packages/resolver/src/adapters/handle-api.ts]:
```suggestion
// https://github.com/koralabs/handles-public-api/blob/fd9d9f2cf3143e317a780b81b22869a755ab6af8/src/models/view/handle.view.model.ts
```

File: apps/wallet-mobile/src/components/TextInput/TextInput.tsx [hunk 19]
```diff
@@ -18,7 +19,7 @@ export type TextInputProps = RNTextInputProps &
   Omit<React.ComponentProps<typeof RNPTextInput>, 'theme'> & {
     containerStyle?: ViewStyle
     renderComponentStyle?: ViewStyle
-    helperText?: string
+    helper?: string | React.ReactNode
```
Comment (banklesss) [inline apps/wallet-mobile/src/components/TextInput/TextInput.tsx:22]:
```suggestion
    helper?: React.ReactNode
```
